File: packages/dart/lib/src/task.dart

import 'workflow.dart';

/// 去掉尾巴上的斜杠，拼路径不出双斜杠。
String _trim(String path) => path.endsWith('/') ? path.substring(0, path.length - 1) : path;

/// 任务聚合：工作流的一次执行实例。
///
/// 指令是跑哪条工作流（[workflowName]，**按名字**引用）与从哪开工（[start]）；
/// 状态是流水（只增不改）、闸门项、产物落点；另带这次执行的运行上下文。
/// 任务是运行数据，不是产物——程序只维护它，不往产物里写字。
///
/// 不可变：[recorded] / [withGates] 都返回新的任务，改动由调用方落盘。
class Task {
  const Task({
    required this.name,
    required this.workflowName,
    this.start = '',
    this.context = const RunContext(),
    this.journal = const [],
    this.gates = const [],
    this.artifacts = const {},
  });

  /// 从任务文件里的字段读出。[name] 由调用方给（文件名即任务名）。
  factory Task.of(String name, Map payload) => Task(
    name: name,
    workflowName: textOf(payload, 'workflow'),
    start: textOf(payload, 'start'),
    context: RunContext.of(payload),
    journal: [
      for (final event in (payload['log'] as List? ?? const []).cast<Map>())
        JournalEvent.of(event),
    ],
    gates: [
      for (final note in payload['gates'] as List? ?? const [])
        if (note is String) note,
    ],
    artifacts: {
      for (final entry in (payload['artifacts'] as Map? ?? const {}).entries)
        if (entry.key is String && entry.value is String)
          '${entry.key}': '${entry.value}',
    },
  );

  final String name;
  final String workflowName;
  final String start;
  final RunContext context;
  final List<JournalEvent> journal;

  /// 等人拍板的事项。
  final List<String> gates;

  /// 这次执行往哪写产物（声明写成什么就是什么，相对工作区根）。
  final Map<String, String> artifacts;

  /// 这种产物声明了往哪写；没声明给 null。
  String? declared(String kind) {
    final written = artifacts[kind]?.trim() ?? '';
    return written.isEmpty ? null : written;
  }

  /// 这次执行往哪写这种产物（规范「任务 / 语法」里的落点）。
  ///
  /// 声明了按声明的（相对工作区根）；没声明落数据仓的
  /// `artifacts/<种类>/<任务名>.md`；流水是任务文件本身。
  String artifact(String kind, RunContext context) {
    final data = _trim(context.data);
    if (kind == 'log') return '$data/tasks/$name.yaml';
    final written = declared(kind);
    if (written != null) {
      return written.startsWith('/') ? written : '${_trim(context.root)}/$written';
    }
    return '$data/artifacts/$kind/$name.md';
  }

  /// 走过哪几步。
  ///
  /// 带后缀的流水（`·审` / `·判`）给这一步**投票**，不带后缀的（重走一遍）
  /// **把结论从头算**；工作流上没有的步骤名不算数。
  List<String> doneSteps(Workflow workflow) {
    final names = workflow.stepNames;
    final verdict = <String, bool>{};
    final order = <String>[];
    for (final event in journal) {
      final at = event.step.indexOf('·');
      final step = at < 0 ? event.step : event.step.substring(0, at);
      if (!names.contains(step)) continue;
      if (verdict.containsKey(step)) {
        verdict[step] = at >= 0 ? (verdict[step]! && event.ok) : event.ok;
      } else {
        verdict[step] = event.ok;
        order.add(step);
      }
    }
    return order.where((step) => verdict[step] == true).toList();
  }

  /// 第一个没走到的步骤（按工作流里的顺序）。
  String? nextStep(Workflow workflow) {
    final finished = doneSteps(workflow);
    for (final name in workflow.stepNames) {
      if (!finished.contains(name)) return name;
    }
    return null;
  }

  /// 状态行：下一步是谁，或者都走过了。
  String stateLine(Workflow workflow) {
    final names = workflow.stepNames;
    if (names.isEmpty) {
      return '这条工作流没有步骤——在 workflows/$workflowName.yaml 的 steps 里写步骤';
    }
    final next = nextStep(workflow);
    return next == null ? '${names.length} 个步骤都走过了' : '下一步：$next';
  }

  /// 记一笔流水：流水只增不改，所以返回新的任务。
  Task recorded({
    required String at,
    required String step,
    required String detail,
    required bool ok,
  }) => Task(
    name: name,
    workflowName: workflowName,
    start: start,
    context: context,
    journal: [
      ...journal,
      JournalEvent(at: at, step: step, detail: detail, ok: ok),
    ],
    gates: gates,
    artifacts: artifacts,
  );

  /// 记下闸门项；已有的不重复记。同样返回新的任务。
  Task withGates(List<String> notes) {
    final all = [...gates];
    for (final note in notes) {
      if (!all.contains(note)) all.add(note);
    }
    return Task(
      name: name,
      workflowName: workflowName,
      start: start,
      context: context,
      journal: journal,
      gates: all,
      artifacts: artifacts,
    );
  }

  Map<String, Object?> toMap() => {
    'name': name,
    'start': start,
    'workflow': workflowName,
    'log': [for (final event in journal) event.toMap()],
    'gates': gates,
    'artifacts': artifacts,
    ...context.toMap(),
  };
}

/// 流水里的一条：什么时候、哪一步、一句话、过没过。
class JournalEvent {
  const JournalEvent({
    required this.at,
    required this.step,
    required this.detail,
    required this.ok,
  });

  factory JournalEvent.of(Map value) => JournalEvent(
    at: textOf(value, 'at'),
    step: textOf(value, 'step'),
    detail: textOf(value, 'detail'),
    ok: value['ok'] == true,
  );

  final String at;
  final String step;
  final String detail;
  final bool ok;

  Map<String, Object?> toMap() => {
    'at': at,
    'step': step,
    'detail': detail,
    'ok': ok,
  };
}

/// 这次执行自带的运行上下文：工作区根、数据仓、工作流目录。
///
/// 三个字段怎么读、怎么写由各自的包定（工具箱只管托着它们）。
class RunContext {
  const RunContext({this.root = '', this.data = '', this.workflows = ''});

  factory RunContext.of(Map value) => RunContext(
    root: textOf(value, 'root'),
    data: textOf(value, 'data'),
    workflows: textOf(value, 'workflows'),
  );

  final String root;
  final String data;
  final String workflows;

  Map<String, Object?> toMap() => {
    'root': root,
    'data': data,
    'workflows': workflows,
  };
}

File: packages/rust/src/task.rs
//! 任务聚合：工作流的一次执行实例。
//!
//! 指令是跑哪条工作流（`workflow_name`，**按名字**引用）与从哪开工（`start`）；
//! 状态是流水（只增不改）、闸门项、产物落点；另带这次执行的运行上下文。
//! 任务是运行数据，不是产物——程序只维护它，不往产物里写字。
//!
//! 不可变：`recorded` / `with_gates` 都返回新的任务，改动由调用方落盘。

use std::collections::BTreeMap;

use crate::workflow::Workflow;
use crate::workflow::text_of;
use serde_yaml::{Mapping, Value as Yaml};

/// 流水里的一条：什么时候、哪一步、一句话、过没过。
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct JournalEvent {
    pub at: String,
    pub step: String,
    pub detail: String,
    pub ok: bool,
}

impl JournalEvent {
    pub fn of(value: &Yaml) -> JournalEvent {
        JournalEvent {
            at: text_of(value, "at"),
            step: text_of(value, "step"),
            detail: text_of(value, "detail"),
            ok: value.get("ok").and_then(|v| v.as_bool()).unwrap_or(false),
        }
    }

    pub fn to_yaml(&self) -> Yaml {
        let mut map = Mapping::new();
        for (key, value) in [
            ("at", &self.at),
            ("step", &self.step),
            ("detail", &self.detail),
        ] {
            map.insert(
                Yaml::String(key.to_string()),
                Yaml::String(value.to_string()),
            );
        }
        map.insert(Yaml::String("ok".into()), Yaml::Bool(self.ok));
        Yaml::Mapping(map)
    }
}

/// 这次执行自带的运行上下文：工作区根、数据仓、工作流目录。
///
/// 三个字段怎么读、怎么写由各自的包定（工具箱只管托着它们）。
#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub struct RunContext {
    pub root: String,
    pub data: String,
    pub workflows: String,
}

impl RunContext {
    pub fn of(value: &Yaml) -> RunContext {
        RunContext {
            root: text_of(value, "root"),
            data: text_of(value, "data"),
            workflows: text_of(value, "workflows"),
        }
    }

    pub fn to_yaml(&self) -> Yaml {
        let mut map = Mapping::new();
        for (key, value) in [
            ("root", &self.root),
            ("data", &self.data),
            ("workflows", &self.workflows),
        ] {
            map.insert(
                Yaml::String(key.to_string()),
                Yaml::String(value.to_string()),
            );
        }
        Yaml::Mapping(map)
    }
}

/// 任务聚合。
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Task {
    pub name: String,
    pub workflow_name: String,
    pub start: String,
    pub context: RunContext,
    pub journal: Vec<JournalEvent>,
    /// 等人拍板的事项。
    pub gates: Vec<String>,
    /// 这次执行往哪写产物（声明写成什么就是什么，相对工作区根）。
    pub products: BTreeMap<String, String>,
}

impl Task {
    /// 从任务文件里的字段读出；`name` 由调用方给（文件名即任务名）。
    pub fn of(name: &str, payload: &Yaml) -> Task {
        let journal = payload
            .get("log")
            .and_then(|v| v.as_sequence())
            .map(|items| items.iter().map(JournalEvent::of).collect())
            .unwrap_or_default();
        let gates = payload
            .get("gates")
            .and_then(|v| v.as_sequence())
            .map(|items| {
                items
                    .iter()
                    .filter_map(|item| item.as_str().map(|text| text.to_string()))
                    .collect()
            })
            .unwrap_or_default();
        let products = payload
            .get("products")
            .and_then(|v| v.as_mapping())
            .map(|mapping| {
                mapping
                    .iter()
                    .filter_map(|(key, value)| {
                        Some((key.as_str()?.to_string(), value.as_str()?.to_string()))
                    })
                    .collect()
            })
            .unwrap_or_default();
        Task {
            name: name.to_string(),
            workflow_name: text_of(payload, "workflow"),
            start: text_of(payload, "start"),
            context: RunContext::of(payload),
            journal,
            gates,
            products,
        }
    }

    /// 这种产物声明了往哪写；没声明给 `None`（落哪是各自包的事）。
    pub fn product(&self, kind: &str) -> Option<String> {
        let written = self.products.get(kind)?.trim();
        if written.is_empty() {
            None
        } else {
            Some(written.to_string())
        }
    }

    /// 走过哪几步。
    ///
    /// 带后缀的流水（`·审` / `·判`）给这一步**投票**，不带后缀的（重走一遍）
    /// **把结论从头算**；工作流上没有的步骤名不算数。
    pub fn done_steps(&self, workflow: &Workflow) -> Vec<String> {
        let names = workflow.step_names();
        let mut verdict: Vec<(String, bool)> = Vec::new();
        for event in &self.journal {
            let (step, extra) = match event.step.split_once('·') {
                Some((step, rest)) => (step.to_string(), Some(rest.to_string())),
                None => (event.step.clone(), None),
            };
            if !names.contains(&step) {
                continue;
            }
            match verdict.iter_mut().find(|(name, _)| *name == step) {
                Some((_, last)) => {
                    if extra.is_some() {
                        *last = *last && event.ok;
                    } else {
                        *last = event.ok;
                    }
                }
                None => verdict.push((step, event.ok)),
            }
        }
        verdict
            .into_iter()
            .filter(|(_, ok)| *ok)
            .map(|(name, _)| name)
            .collect()
    }

    /// 第一个没走到的步骤（按工作流里的顺序）。
    pub fn next_step(&self, workflow: &Workflow) -> Option<String> {
        let finished = self.done_steps(workflow);
        workflow
            .step_names()
            .into_iter()
            .find(|name| !finished.contains(name))
    }

    /// 状态行：下一步是谁，或者都走过了。
    pub fn state_line(&self, workflow: &Workflow) -> String {
        let names = workflow.step_names();
        if names.is_empty() {
            return format!(
                "这条工作流没有步骤——在 workflows/{}.yaml 的 steps 里写步骤",
                self.workflow_name
            );
        }
        match self.next_step(workflow) {
            Some(step) => format!("下一步：{step}"),
            None => format!("{} 个步骤都走过了", names.len()),
        }
    }

    /// 记一笔流水：流水只增不改，所以返回新的任务。
    pub fn recorded(&self, at: &str, step: &str, detail: &str, ok: bool) -> Task {
        let mut task = self.clone();
        task.journal.push(JournalEvent {
            at: at.to_string(),
            step: step.to_string(),
            detail: detail.to_string(),
            ok,
        });
        task
    }

    /// 记下闸门项；已有的不重复记。同样返回新的任务。
    pub fn with_gates(&self, notes: &[String]) -> Task {
        let mut task = self.clone();
        for note in notes {
            if !task.gates.contains(note) {
                task.gates.push(note.clone());
            }
        }
        task
    }

    /// 写回任务文件的字段形状。
    pub fn to_yaml(&self) -> Yaml {
        let mut map = Mapping::new();
        map.insert(Yaml::String("name".into()), Yaml::String(self.name.clone()));
        map.insert(
            Yaml::String("start".into()),
            Yaml::String(self.start.clone()),
        );
        map.insert(
            Yaml::String("workflow".into()),
            Yaml::String(self.workflow_name.clone()),
        );
        map.insert(
            Yaml::String("log".into()),
            Yaml::Sequence(self.journal.iter().map(JournalEvent::to_yaml).collect()),
        );
        map.insert(
            Yaml::String("gates".into()),
            Yaml::Sequence(
                self.gates
                    .iter()
                    .map(|note| Yaml::String(note.clone()))
                    .collect(),
            ),
        );
        let mut products = Mapping::new();
        for (key, value) in &self.products {
            products.insert(Yaml::String(key.clone()), Yaml::String(value.clone()));
        }
        map.insert(Yaml::String("products".into()), Yaml::Mapping(products));
        if let Yaml::Mapping(context) = self.context.to_yaml() {
            for (key, value) in context {
                map.insert(key, value);
            }
        }
        Yaml::Mapping(map)
    }
}
